File: rfid_reader/lib/src/rfid_reader.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:my_rfid_reader/my_rfid_reader_mixin.dart';
import 'package:my_rfid_reader/my_rfid_reader_util.dart';

class RfidReader extends StatefulWidget {
  final Control? parent;
  final Control control;
  final FletControlBackend backend;

  const RfidReader(
      {super.key, this.parent, required this.control, required this.backend});
  @override
  _RfidReader createState() => _RfidReader();
}

class _RfidReader extends State<RfidReader> with MyRfidReaderMixin<RfidReader>{
  late TextEditingController _controller;
  String _string = "";
  MyRfidReaderUtil myRfidReaderUtil = MyRfidReaderUtil();

  @override
  Widget build(BuildContext context) {
    var message_tag = widget.control.attrString("message_tag");
    var start_connect = widget.control.attrBool("start_connect")!;
    // var close_connect = widget.control.attrBool("close_connect")!;
    var turn_on_power = widget.control.attrBool("turn_on_power")!;
    var turn_off_power = widget.control.attrBool("turn_off_power")!;
    var start_reader = widget.control.attrBool("start_reader")!;
    var start_reader_epc = widget.control.attrBool("start_reader_epc")!;
    var start_write_epc = widget.control.attrBool("start_write_epc")!;
    var epc_data = widget.control.attrString("epc_data");
    var epc_data_area = widget.control.attrInt("epc_data_area");
    var connect_message = widget.control.attrString("connect_message");
    var power_message = widget.control.attrString("power_message");
    var reader_operation_message = widget.control.attrString("reader_operation_message");
    var epc_messages = widget.control.attrString("epc_messages");
    var write_epc_message = widget.control.attrString("write_epc_message");


    if (start_connect) {     // 开始扫描标志为True时向Android端发送对应信息
      myRfidReaderUtil.sendMessageToAndroid("startConnect", true);
      setState(() {
        /*
          发送完消息后将开始扫描标志置为“false”（目前只掌握更新字符串属性方式，
          不过影响不大python控件那边布尔型属性可以接受字符串）
          下方多个标志一样的处理方式
        */
        widget.backend.updateControlState(widget.control.id, {"start_connect": "false"});
      });
    }
    if (turn_on_power)  {
      myRfidReaderUtil.sendMessageToAndroid("turnOnPower", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"turn_on_power": "false"});
      });
    }
    if (turn_off_power) {
      myRfidReaderUtil.sendMessageToAndroid("turnOffPower", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"turn_off_power": "false"});
      });
    }
    if (start_reader) {
      myRfidReaderUtil.sendMessageToAndroid("startReader", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_reader": "false"});
      });
    }
    if (start_reader_epc) {
      myRfidReaderUtil.sendMessageToAndroid("startReaderEpc", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_reader_epc": "false"});
      });
    }
    if (start_write_epc && epc_data != "") {
      String write_epc_data = "";
      if (epc_data != null) {
        write_epc_data = epc_data + "&" + epc_data_area.toString();
      }
      myRfidReaderUtil.sendMessageToAndroid("writeEpcData", write_epc_data);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_write_epc": "false"});
      });
    }
    if (_string != "") {
      _controller.text = _string;
    }
    Widget bluetooth_control = TextField(
        controller: _controller,
        decoration: const InputDecoration(
          border: InputBorder.none, //无边框
        )
    );
    return constrainedControl(
        context, bluetooth_control, widget.parent, widget.control);
  }

  @override
  Future<void> listenerAndroidHandle(dynamic message) async {
    // TODO: implement listenerAndroidHandle
    // 连接结果（连接成功、连接失败、断开连接）消息样式为   连接成功>>>（设备名称）
    var connectMessage = message["connectMessage"];
    if (connectMessage != null) {
      widget.backend.updateControlState(widget.control.id, {"connect_message": connectMessage});
      String eventString = connectMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "connect_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 读写器读卡操作信息
    var readerOperationMessage = message["readerOperationMessage"];
    if (readerOperationMessage != null) {
      widget.backend.updateControlState(widget.control.id, {"reader_operation_message": readerOperationMessage});
      String eventString = readerOperationMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "reader_operation_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 读取epc数据
    var epcMessages = message["epcMessages"];
    if (epcMessages != null) {
      String eventString = "";
      for (String epcMessage in epcMessages)  {
        eventString += "$epcMessage&";
      }
      widget.backend.updateControlState(widget.control.id, {"epc_messages": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "epc_messages"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 上下电信息
    var powerMessage = message["powerMessage"];
    if (powerMessage != null) {
      String eventString = powerMessage;
      widget.backend.updateControlState(widget.control.id, {"power_message": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "power_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }

    // 写卡信息
    var writeEpcMessage = message["writeEpcMessage"];
    if (writeEpcMessage != null) {
      String eventString = writeEpcMessage;
      widget.backend.updateControlState(widget.control.id, {"write_epc_message": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "write_epc_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
  }
}

File: python/controls/rfid_reader.py
from flet_core.constrained_control import ConstrainedControl
from typing import Optional
import binascii

class RfidReader(ConstrainedControl):
    """
    RfidReader 控件
    """

    def __init__(
        self,
        message_tag: Optional[str] = "",
        epc_data: Optional[str] = "",
        epc_data_area: Optional[int] = 0,
        start_connect: Optional[bool] = False,
        # close_connect: Optional[bool] = False,
        turn_on_power: Optional[bool] = False,
        turn_off_power: Optional[bool] = False,
        start_reader: Optional[bool] = False,
        start_reader_epc: Optional[bool] = False,
        start_write_epc: Optional[bool] = False,
        connect_message: Optional[str] = "",
        power_message: Optional[str] = "",
        epc_messages: Optional[str] = "",
        write_epc_message: Optional[str] = "",
        reader_operation_message: Optional[str] = "",
        on_listener = None,
    ):
        ConstrainedControl.__init__(self)
        self.start_connect = start_connect     # 设备连接标志
        # self.close_connect = close_connect     # 断开设备连接标志
        self.turn_on_power = turn_on_power     # 设备上电标志
        self.turn_off_power = turn_off_power   # 设备下电标志
        self.start_reader = start_reader   # 启动读写器读卡标志
        self.start_reader_epc = start_reader_epc    # 读取epc标签数据标志
        self.connect_message = connect_message    # 连接设备时返回的提示信息
        self.power_message = power_message   # 设备上下电信息
        self.epc_messages = epc_messages    # 单次盘点所获取的epc标签数据
        self.reader_operation_message = reader_operation_message    # 启动读写器读卡返回的信息
        self.on_listener = on_listener     # 用于监听Android端发送过来的信息事件方法
        self.message_tag = message_tag    # 用于甄别Android端发送过来的信息是何种类型
        self.epc_data = epc_data           # 需要向标签写入的数据（仅限于十六进制）在输入的时候进行限制较为简单
        self.epc_data_area = epc_data_area     #  写入数据的区域
        self.start_write_epc = start_write_epc    #  进行写卡的标志
        self.write_epc_emssage = write_epc_message   #  写卡时返回过来的信息

    def _get_control_name(self) -> str:
        return "rfid_reader"


    
    @property
    def start_connect(self):
        return self._get_attr("start_connect")

    @start_connect.setter
    def start_connect(self, value):
        self._set_attr("start_connect", value)

    # @property
    # def close_connect(self):
    #     return self._get_attr("close_connect")

    # @close_connect.setter
    # def close_connect(self, value):
    #     self._set_attr("close_connect", value)
    
    @property
    def turn_on_power(self):
        return self._get_attr("turn_on_power")

    @turn_on_power.setter
    def turn_on_power(self, value):
        self._set_attr("turn_on_power", value)
    
    @property
    def turn_off_power(self):
        return self._get_attr("turn_off_power")

    @turn_off_power.setter
    def turn_off_power(self, value):
        self._set_attr("turn_off_power", value)

    @property
    def start_reader(self):
        return self._get_attr("start_reader")

    @start_reader.setter
    def start_reader(self, value):
        self._set_attr("start_reader", value)

    @property
    def start_reader_epc(self):
        return self._get_attr("start_reader_epc")

    @start_reader_epc.setter
    def start_reader_epc(self, value):
        self._set_attr("start_reader_epc", value)

    @property
    def connect_message(self):
        return self._get_attr("connect_message")

    @connect_message.setter
    def connect_message(self, value):
        self._set_attr("connect_message", value)

    @property
    def epc_messages(self):
        return self._get_attr("epc_messages")

    @epc_messages.setter
    def epc_messages(self, value):
        self._set_attr("epc_messages", value)
    
    @property
    def power_message(self):
        return self._get_attr("power_message")

    @power_message.setter
    def power_message(self, value):
        self._set_attr("power_message", value)
    
    @property
    def reader_operation_message(self):
        return self._get_attr("reader_operation_message")

    @reader_operation_message.setter
    def reader_operation_message(self, value):
        self._set_attr("reader_operation_message", value)

    @property
    def on_listener(self):
        return self._get_event_handler("on_listener")

    @on_listener.setter
    def on_listener(self, handler):
        self._add_event_handler("on_listener", handler)
        # self._set_attr("start_listener", True if handler is not None else None)
    @property
    def message_tag(self):
        return self._get_attr("message_tag")

    @message_tag.setter
    def message_tag(self, value):
        self._set_attr("message_tag", value)

    @property
    def epc_data(self):
        return self._get_attr("epc_data")

    @epc_data.setter
    def epc_data(self, value):
        self._set_attr("epc_data", value)
    
    @property
    def start_write_epc(self):
        return self._get_attr("start_write_epc")

    @start_write_epc.setter
    def start_write_epc(self, value):
        self._set_attr("start_write_epc", value)
    
    @property
    def epc_data_area(self):
        return self._get_attr("epc_data_area")

    @epc_data_area.setter
    def epc_data_area(self, value):
        self._set_attr("epc_data_area", value)

    @property
    def write_epc_message(self):
        return self._get_attr("write_epc_message")

    @write_epc_message.setter
    def write_epc_message(self, value):
        self._set_attr("write_epc_message", value)
    
    def connect(self):
        self._set_attr("start_connect", True)
    
    def turn_on(self):
        self._set_attr("turn_on_power", True)
    
    def turn_off(self):
        self._set_attr("turn_off_power", True)
    
    def reader(self):
        self._set_attr("start_reader", True)
    
    def reader_epc(self):
        self._set_attr("start_reader_epc", True)
    
    def write_epc(self, epc_data_area, epc_data):
        epc_data = binascii.hexlify(epc_data.encode("utf-8")).decode("utf-8")    #  转换为十六进制
        self._set_attr("epc_data", epc_data)
        self._set_attr("epc_data_area", epc_data_area)
        self._set_attr("start_write_epc", True)


